### ISSUE TITLE
Service discovery and ip addresses

### ISSUE THREAD

**Issue by miracle2k** [flynn-archive/sdutil#4]
I want a service in a docker container to register itself with discoverd.

I'm a bit confused about this, so this may partly be a question.

I was imaging a container would register itself with `my-docker0-ip:port-my-service-runs-on`. Other containers can now connect to it, done. But this seems not how things were designed.

One, the APIs for discoverd and sdutil seem to have a preference for registering a port only. I find this strange because it would seem this only works when the container port is mapped to the host, _and_ the consuming client knows which host it is.

Two, say I do set a host. In slugrunner I need to specify a SD_HOST env. I obviously do not know the IP of the container at this point, so I'd have to specify the host address and, again, it would only work if I expose the container port on the host (and a fixed port, too; one assigned by docker the container wouldn't know about).

Unless I am missing how this is supposed to work, what I would like to add to sdutil a mode where it'll use the ip of an interface:

```
 sdutil exec -i eth0 example-server:9999
 (register example-server with 172.17.42.28:9999)
```

Further, I'd like slugrunner to support this mode.

**Comment by titanous** [comment]
Currently in Flynn `discoverd` knows the IP of the host it is on and there is one `discoverd` per host, which is why we don't usually specify the IP. Also, the container IP isn't really useful in most environments, because it is only usable from other containers on the host.

That being said, I could see this being useful outside of the context of Flynn. I guess we need to parse something in `/proc` to get the IP of the specified interface?

**Comment by miracle2k** [comment]
I see. My setup is different, because all the hosts are all in the same LAN.

So do I get this right, if I want to connect a slugrunner deployed app to strowger, using the Flynn setup, then the slugrunner would register itself with host-ip:port, with the controller on the host choosing a different free host port for each slugrunner instance? Something I could consider.

As for getting the ip address, I believe http://golang.org/pkg/net/#Interface.Addrs can be used.

**Comment by titanous** [comment]
Yeah, in Flynn our `host` service picks free ports and assigns/exposes them to containers. In most public cloud environments and a lot of other places, the docker container IPs are not exposed to the LAN due to strict network policies, so we've standardized on using the host IP for now.

**Comment by progrium** [comment]
Discoverd supports full Go addrs (ip:port), but if you don't specify an ip
the Discoverd daemon adds the public ip it's configured with. Sdutil should
let you specify a host with the -h option unless it was removed .. Looking
up the ip of an interface sounds like a reasonable convenience to live in
sdutil.

On Tuesday, February 4, 2014, Jonathan Rudenberg notifications@github.com
wrote:

> Yeah, in Flynn our host service picks free ports and assigns/exposes them
> to containers. In most public cloud environments and a lot of other places,
> the docker container IPs are not exposed to the LAN due to strict network
> policies, so we've standardized on using the host IP for now.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/flynn/sdutil/issues/4#issuecomment-34080558
> .

## 

Jeff Lindsay
http://progrium.com

**Comment by titanous** [comment]
@miracle2k Are you interested in implementing the `-interface` flag?

**Comment by miracle2k** [comment]
I'm interested, but after #5.

**Comment by titanous** [comment]
Cool, looking forward to it.

**Comment by titanous** [comment]
Closed by 68404831a07357529c78557d39f3a6c18577e1d0.
